tests/integration_tests/test_gym_envs.py

```diff
@@ -0,0 +1,130 @@
+
+    garage_envs = [TfEnv, TheanoEnv]
+
+    def _routine_test_pickling_wrapper(self, garage_env, spec):
+        env = garage_env(env_name=spec.id)
+
+        # Roundtrip serialization
+        env = pickle.loads(pickle.dumps(env))
+
+        # Step and render still work after pickling
+        env.reset()
+        env.step(env.action_space.sample())
+
+        try:
+            env.render()
+        except NotImplementedError:
+            assert spec.id in not_rendering_envs, "It seems environment %s " \
+                    "cannot be rendered. Add it to the corresponding white" \
+                    "list in test_gym_envs.py." % (spec.id)
+
+        # Test serialization once the viewer object has been set in env, but
+        # don't reassign env with the one produced with pickle to avoid
+        # losing the reference to the viewer object which is required in
+        # env.close()
+        pickled_env = pickle.loads(pickle.dumps(env))
+        assert pickled_env
+
+        try:
+            env.close()
+        except AttributeError:
+            assert spec.id in not_closing_envs, "It seems environment %s " \
+                    "cannot be closed. Add it to the corresponding white" \
+                    "list in test_gym_envs.py." % (spec.id)
+
+    @params(*list(product(garage_envs, list(gym.envs.registry.all()))))
+    def test_pickling_wrapper(self, garage_env, spec):
+        # Skip Atari environments
+        if spec._entry_point == "gym.envs.atari:AtariEnv":
+            return
+        process = Process(
+            target=self._routine_test_pickling_wrapper,
+            args=(
+                garage_env,
+                spec,
+            ))
+        process.start()
+        process.join()
+        assert not process.exitcode
+
+    def _routine_test_step_and_render(self, garage_env, spec):
+        env = garage_env(env_name=spec.id)
+
+        # Step and render work
+        env.reset()
+        done = False
+        for _ in range(10):
+            _, _, done, _ = env.step(env.action_space.sample())
+            try:
+                env.render()
+            except NotImplementedError:
+                assert spec.id in not_rendering_envs, "It seems environment " \
+                        "%s cannot be rendered. Add it to the corresponding " \
+                        "white list in test_gym_envs.py." % (spec.id)
+            if done:
+                break
+
+        try:
+            env.close()
+        except AttributeError:
+            assert spec.id in not_closing_envs, "It seems environment %s " \
+                    "cannot be closed. Add it to the corresponding white" \
+                    "list in test_gym_envs.py." % (spec.id)
+
+    @params(*list(product(garage_envs, list(gym.envs.registry.all()))))
+    def test_step_and_render(self, garage_env, spec):
+        # Skip Atari environments
+        if spec._entry_point == "gym.envs.atari:AtariEnv":
+            return
+        process = Process(
```

Comment:
Please don't launch new processes in tests.

Reply:
Why did you launch a process?

Reply:
I'm launching processes here due to MjViewers. By running this test in my computer, the windows for the MjViewers aren't closed with env.close(), but until the parent process finishes, and since there's a lot of MuJoCo related environments, there are a lot of windows rendering MuJoCo simulations by the middle of the test which leads to shortage of RAM and sometimes termination of the test process to keep the computer alive.
By adding a process for each MjViewer I'm able to close them right after the env is closed, but I agree that this may not the best way to solve this problem, and it also seems to be generating problems in Travis, specifically this error:
```
XIO:  fatal IO error 11 (Resource temporarily unavailable) on X server ":0"
      after 81 requests (81 known processed) with 0 events remaining.
```
Maybe there's a way to detect and close this windows with python.

Reply:
Jonathon and I have this problem too, both on Linux and Mac. It's a bug in mujoco_py. However it's only a problem when testing--because most people don't launch 100+ viewer windows.

It's bad to launch a new process when you test, because it messes with the test runner, and also is not faithful to how the code is executed.

I deal with it when testing locally by running tests with xvfb-run, which makes the windows go to a virtual display.

`xvfb-run --server-args="-screen 0 1024x768x16 +iglx" nose2 ...`